tests/e2e-evm/test/abi_basic.test.ts

```diff
@@ -0,0 +1,162 @@
+import hre from "hardhat";
+import type { ArtifactsMap } from "hardhat/types/artifacts";
+import type { PublicClient, WalletClient, ContractName } from "@nomicfoundation/hardhat-viem/types";
+import { expect } from "chai";
+import { Address, toFunctionSelector, toFunctionSignature, concat } from "viem";
+import { getAbiFallbackFunction, getAbiReceiveFunction } from "./helpers/abi";
+import { whaleAddress } from "./addresses";
+
+interface ContractTestCase {
+  interface: keyof ArtifactsMap;
+  mock: ContractName<keyof ArtifactsMap>;
+}
+
+// ABI_BasicTests assert ethereum + solidity transaction ABI interactions perform as expected.
+describe("ABI_BasicTests", function () {
+  const testCases = [
+    // Test function modifiers without receive & fallback
+    { interface: "NoopNoReceiveNoFallback", mock: "NoopNoReceiveNoFallbackMock" },
+
+    // Test receive + fallback scenarios
+    { interface: "NoopReceiveNoFallback", mock: "NoopReceiveNoFallbackMock" },
+    { interface: "NoopReceivePayableFallback", mock: "NoopReceivePayableFallbackMock" },
+    { interface: "NoopReceiveNonpayableFallback", mock: "NoopReceiveNonpayableFallbackMock" },
+
+    // Test no receive + fallback scenarios
+    { interface: "NoopNoReceivePayableFallback", mock: "NoopNoReceivePayableFallbackMock" },
+    { interface: "NoopNoReceiveNonpayableFallback", mock: "NoopNoReceiveNonpayableFallbackMock" },
+  ] as ContractTestCase[];
+
+  //
+  // Client + Wallet Setup
+  //
+  let publicClient: PublicClient;
+  let walletClient: WalletClient;
+  before("setup clients", async function () {
+    publicClient = await hre.viem.getPublicClient();
+    walletClient = await hre.viem.getWalletClient(whaleAddress);
+  });
+
+  //
+  // Test each function defined in the interface ABI
+  //
+  // Only payable functions may be sent value
+  // Any function (payable, non-payable, view, pure) can be called via a transaction
+  // All functions can be provided calldata regardless of their arguments
+  for (const tc of testCases) {
+    describe(tc.interface, function () {
+      const abi = hre.artifacts.readArtifactSync(tc.interface).abi;
+      const receiveFunction = getAbiReceiveFunction(abi);
+      const fallbackFunction = getAbiFallbackFunction(abi);
+
+      let mockAddress: Address;
+      before("deploy mock", async function () {
+        mockAddress = (await hre.viem.deployContract(tc.mock)).address;
+      });
+
+      describe("State", function () {
+        it("has code set", async function () {
+          const code = await publicClient.getCode({ address: mockAddress });
+          expect(code).to.not.equal(0);
```

Comment:
can you confirm that an EOA or never-used address has code `0`? i am not 💯 familiar with these operations by i just want to confirm `getCode` isn't returning a string like `"0x"` (which also is not equal to the number `0`)

Reply:
You are probably correct here, I believe this is suppose to be checked against undefined -- Viem converts the "0x" response to undefined, I think I confused this assetion with balance

Reply:
Could we compare it with an code match from the `deployedBytecode` in the abi instead of a non-match?